src/realm/parser/driver.hpp

```diff
@@ -19,59 +21,64 @@ namespace realm {
 
 namespace query_parser {
 
+class NodeVisitor;
+
 class ParserNode {
 public:
     virtual ~ParserNode();
+    virtual void accept(NodeVisitor& visitor) = 0;
 };
 
 class AtomPredNode : public ParserNode {
 public:
     ~AtomPredNode() override;
-    virtual Query visit(ParserDriver*) = 0;
 };
 
-class AndNode : public ParserNode {
+class AndNode : public AtomPredNode {
 public:
```

Comment:
I get nervous when I see these changes, I have not yet figured out if it is a problem,